instrumentation/jms-1.1/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/jms/JmsMessageConsumerInstrumentation.java

```diff
@@ -49,28 +52,27 @@
   public static class ConsumerAdvice {
 
     @Advice.OnMethodEnter
-    public static long onEnter() {
-      return System.currentTimeMillis();
+    public static Instant onEnter() {
+      return Instant.now();
     }
 
     @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
     public static void stopSpan(
-        @Advice.Enter long startTime,
+        @Advice.Enter Instant startTime,
         @Advice.Return Message message,
         @Advice.Thrown Throwable throwable) {
-      MessageDestination destination;
       if (message == null) {
         // Do not create span when no message is received
         return;
       }
-      destination = tracer().extractDestination(message, null);
 
-      Context context = tracer().startConsumerSpan(destination, "receive", message, startTime);
+      Context parentContext = Java8BytecodeBridge.currentContext();
+      MessageWithDestination request =
+          MessageWithDestination.create(message, MessageOperation.receive, null);
 
-      if (throwable != null) {
-        tracer().endExceptionally(context, throwable);
-      } else {
-        tracer().end(context);
+      if (consumerInstrumenter().shouldStart(parentContext, request)) {
+        Context context = consumerInstrumenter().start(parentContext, request, startTime);
```

Comment:
Any thoughts on adding `startTime` to `MessageWithDestination` and adding `StartTimeExtractor` to read it from something? If using Armeria's start time, which I would if we do end up needing this knob more generally :), it would just be reading the time off the request object. Current approach might be more general though, wondering what you think. I think `StartTimeExtractor` hides the complexity more than adding to our core API, something to consider if this is extremely rarely needed.

Reply:
Oh, that's much better than passing the argument -- will do.

Reply:
And I can add an `EndTimeExtractor` too while we're working on it - the `end()` method already has too many params.